File: apps/python/firstbell/tools/gates/frame-sweep.mjs
/* Does any pose the drag allows put ink on the edge of the board's own canvas?
 *
 * Not a gate; run by hand while tuning the framing. Sweeps yaw right round and pitch across
 * its clamp, photographs the canvas at each pose, and writes one PNG per pose for
 * `frame-sweep.py` to count edge pixels on. Pixels rather than bounding boxes: see the
 * comment on `pose` in morning.js for why the box answer is wrong here.
 */
import { createServer } from "node:http";
import { readFile } from "node:fs/promises";
import { existsSync, mkdirSync } from "node:fs";
import { extname, join, resolve } from "node:path";
import puppeteer from "puppeteer-core";
import { listenSafely } from "./safe-port.mjs";

const MIME = { ".html": "text/html", ".js": "text/javascript", ".css": "text/css",
  ".json": "application/json", ".svg": "image/svg+xml", ".png": "image/png",
  ".mp3": "audio/mpeg", ".woff2": "font/woff2", ".ico": "image/x-icon", ".txt": "text/plain" };
const ROOT = resolve(process.argv[2] || "out");
const SHOTS = resolve(process.argv[3] || "sweep");
const CASES = JSON.parse(process.argv[4]);
const YAWS = Number(process.argv[5] || 16);
const CHROME = [process.env.CHROME_PATH, "C:/Program Files/Google/Chrome/Application/chrome.exe",
  "C:/Program Files (x86)/Google/Chrome/Application/chrome.exe"].find((p) => p && existsSync(p));

mkdirSync(SHOTS, { recursive: true });
const server = createServer(async (req, res) => {
  const rel = decodeURIComponent(req.url.split("?")[0]);
  try {
    const body = await readFile(join(ROOT, rel === "/" ? "index.html" : rel));
    res.writeHead(200, { "content-type": MIME[extname(rel).toLowerCase()] || "application/octet-stream" }).end(body);
  } catch { res.writeHead(404).end("no"); }
});
const { port } = await listenSafely(server);
const browser = await puppeteer.launch({ executablePath: CHROME, headless: "new",
  args: ["--use-gl=angle", "--use-angle=swiftshader", "--enable-unsafe-swiftshader"] });

let poses = 0;
for (const [label, path_, stage, width] of CASES) {
  const key = stage === "[data-mrn-stage]" ? "__morning" : "__cutoff";
  const page = await browser.newPage();
  await page.setViewport({ width, height: 900, deviceScaleFactor: 1 });
  await page.goto(`http://127.0.0.1:${port}${path_}`, { waitUntil: "load" });
  const range = await page.evaluate(async (k) => {
    for (let i = 0; i < 120 && !window[k]; i += 1) await new Promise((r) => setTimeout(r, 100));
    return window[k] ? window[k].pitchRange : null;
  }, key);
  if (!range) { console.log(`${label}: no board`); await page.close(); continue; }
  const el = await page.$(`${stage} canvas`);
  const [lo, hi] = range;
  const pitches = [lo, lo * 0.5, 0, hi * 0.5, hi];
  for (let i = 0; i < YAWS; i += 1) {
    const turn = (i * 2 * Math.PI) / YAWS;
    for (const pitch of pitches) {
      await page.evaluate((k, t, p) => { window[k].pose(t, p); }, key, turn, pitch);
      await new Promise((r) => setTimeout(r, 90));
      await el.screenshot({ path: join(SHOTS,
        `${label.replace(/[^a-z0-9]+/gi, "-")}__y${i}__p${pitch.toFixed(2)}.png`) });
      poses += 1;
    }
  }
  console.log(`${label}: ${YAWS * pitches.length} poses`);
  await page.close();
}
await browser.close();
server.close();
console.log(`${poses} poses written to ${SHOTS}`);

File: apps/python/firstbell/tools/gates/frame-sweep.py
"""Count edge pixels in every pose `frame-sweep.mjs` photographed.

The ground is read off the corner of each shot rather than assumed, because the canvas is
transparent and what sits behind it is the page's own paper. A pose with any ink in the
outer band is a pose a reader can drag the board out of the frame in.
"""
from __future__ import annotations

import sys
from pathlib import Path

from PIL import Image

BAND = 3
TOL = 8


def edge_ink(path: Path) -> dict[str, int]:
    im = Image.open(path).convert("RGB")
    w, h = im.size
    px = im.load()
    ground = px[0, 0]

    def ink(x: int, y: int) -> bool:
        p = px[x, y]
        return max(abs(p[i] - ground[i]) for i in range(3)) > TOL

    return {
        "top": sum(ink(x, b) for b in range(BAND) for x in range(w)),
        "bottom": sum(ink(x, h - 1 - b) for b in range(BAND) for x in range(w)),
        "left": sum(ink(b, y) for b in range(BAND) for y in range(h)),
        "right": sum(ink(w - 1 - b, y) for b in range(BAND) for y in range(h)),
    }


def main() -> int:
    shots = sorted(Path(sys.argv[1]).glob("*.png"))
    if not shots:
        print("no shots to read")
        return 1
    bad = []
    for shot in shots:
        hit = edge_ink(shot)
        if any(hit.values()):
            bad.append((shot.stem, hit))
    print(f"{len(shots)} poses read")
    for stem, hit in bad[:25]:
        print(f"  CLIPS  {stem}  t/b/l/r {hit['top']}/{hit['bottom']}/{hit['left']}/{hit['right']}")
    if len(bad) > 25:
        print(f"  ... and {len(bad) - 25} more")
    print(f"{len(bad)} of {len(shots)} poses put ink on the canvas edge")
    return 1 if bad else 0


if __name__ == "__main__":
    raise SystemExit(main())
